### DISCUSSION TITLE
Database performance optimization

### DISCUSSION THREAD

**Discussion by lavsurgut** [apache/airflow#26199] [Q&A] [answered]
Dear Airflow Team,
We notice that Airflow is quite "db hungry". E.g. we run 2.3.4 version of Airflow with KubernetesExecutor and Postgres 11 (2VCPU, 3.75GB RAM) - without any additional changes to airflow.cfg - we see that Airflow fires tens of thousands of queries to the metadata database in a matter of hour (even though there is no tasks to run, nothing is in the queue). In addition, when real tasks are executed - Airflow utilizes almost 100% CPU for the whole tasks run duration. Is it something expected or might be something wrong with our setup?

**Comment by potiuk** [comment, marked as answer]
Depends what database you have and DAGs you create. If you have any suspected queries you see that are excessive, let us know, but in general airflow does use a lot of DB queries continuously even if nothing happens. But you might also generate yourself a lot of DB pressure if your dags do not follow Best Practices, specifically top level code: https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code 

Read the docs, apply learnings (my guess you are using variables at top level code or some other code that generates DB queries).

Let us know what you've learned  from it and whether it worked.

**Reply by lavsurgut** [reply]
Yes, we do use Variables in top level code and plan to re-work it - lets see. Thank you very much
